### PR TITLE
Initial subgroup pass.

### PR DESCRIPTION
**Pull request by jbruestle**
Initial version of transform, cost model + tests still missing.

### CODE REVIEW

File: pmlc/dialect/pxa/transforms/BUILD [hunk 1]
```diff
@@ -31,6 +31,7 @@ plaidml_cc_library(
         "resize_tmps.cc",
         "stencil.cc",
         "stencil_gemm.cc",
+	"subgroups.cc",
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/BUILD:34]: whitespace (use buildifier)

**Comment by jbruestle** [comment]
Seems to work when tested by hand, and shouldn't break anything.  Need to backfill tests of various components, but putting in PR now for faster integration.

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        {8, 16}, // Subgroup sizes
+        6,       // Maximum register per thread
+    };
+    SubgroupCostModel cm(params, op);
+    if (cm.bestCost == std::numeric_limits<double>::infinity()) {
+      return;
+    }
+    IVLOG(1, "best plan = " << cm.bestPlan);
+    SubgroupApply(op, cm.bestPlan);
+    // SubgroupApply(op, plan);
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/subgroups.cc:238]: Still need this?

File: pmlc/dialect/pxa/transforms/cache.cc [hunk 3]
```diff
@@ -99,6 +99,68 @@ LogicalResult cacheLoad(AffineParallelOp par, PxaLoadOp load) {
...
+  auto loadMap = convertToValueMap(load.getContext(), rap.outer);
+  IVLOG(2, "Making vector load");
+  auto loadVec = builder.create<PxaVectorLoadOp>(loc, vecType, load.getMemRef(),
+                                                 loadMap.getAffineMap(),
+                                                 loadMap.getOperands());
+  // Make a new load and remove the old one
+  OpBuilder newLoadBuilder(load);
+  // Do an affine apply to get the index
+  auto innerMap = convertToValueMap(par.getContext(), rap.inner);
+  // Extrace the right element of the vector
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/cache.cc:154]: typo: Extract

File: pmlc/dialect/pxa/transforms/cache.cc [hunk 4]
```diff
@@ -139,10 +201,12 @@ LogicalResult cacheReduce(AffineParallelOp par, PxaReduceOp reduce) {
   auto eltType = reduce.getMemRefType().getElementType();
   auto type = MemRefType::get(rap.innerCount, eltType);
   auto localBuf = builder.create<AllocOp>(loc, type);
-  // Clear it to the reduction identity
-  auto ident = createIdentity(builder, loc, reduce.agg(), eltType);
-  auto initBuf = createInitLoop(builder, loc, localBuf, ident);
-
+  // If it's not an assign, clear it to the reduction dentity
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/cache.cc:204]: Typo: identity

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+struct SubgroupParams {
+  SmallVector<int64_t, 4> subgroupSizes;
+  int64_t maxRegsPerThread;
+};
+
+struct SubgroupCostModel {
+  SubgroupCostModel(const SubgroupParams &params, AffineParallelOp op)
+      : params(params), op(op) {
+    bestCost = std::numeric_limits<double>::infinity();
+    // Verify that we have only 1 results
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/subgroups.cc:61]: Can you say why we care about 1 result (it's easy to see that the code checks for results != 1)?

File: pmlc/dialect/pxa/transforms/vectorize.cc [hunk 6]
```diff
@@ -289,4 +290,102 @@ std::unique_ptr<mlir::Pass> createVectorizeExamplePass() {
...
+      newShape.push_back(mshape[i]);
+    }
+  }
+  // Make the new type
+  auto newType = MemRefType::get(
+      newShape, VectorType::get({vecSize}, mtype.getElementType()));
+  // Replace the alloc
+  auto newOp = replaceOp<AllocOp>(op, newType);
+  // Walk over the uses and update them all
+  auto cur_use = IndirectUsesIterator(newOp);
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/vectorize.cc:372]: nit: curUse

File: pmlc/dialect/pxa/transforms/subgroups.cc [hunk 7]
```diff
@@ -58,7 +58,7 @@ struct SubgroupCostModel {
   SubgroupCostModel(const SubgroupParams &params, AffineParallelOp op)
       : params(params), op(op) {
     bestCost = std::numeric_limits<double>::infinity();
-    // Verify that we have only 1 results
+    // Tile accumulations only works on parallel loops with a singe result
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/subgroups.cc:61]: typo: single

**Pull request merged**